src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+        /// <remarks>
+        /// The parameter lspSnapshot refers to the snapshot used when the LSP document symbol request was made
+        /// to obtain the symbol and currentSnapshot refers to the latest snapshot in the editor. These parameters
+        /// are required to obtain the latest DocumentSymbolViewModel range positions in the new snapshot.
+        /// </remarks>
+        public static void SelectDocumentNode(
+            ImmutableArray<DocumentSymbolViewModel> symbolTreeItemsSource,
+            ITextSnapshot currentSnapshot,
+            ITextSnapshot lspSnapshot,
+            int caretPosition)
+        {
+            LspSnapshot = lspSnapshot;
+            CurrentSnapshot = currentSnapshot;
+            CaretPosition = caretPosition;
+            IsSnapshotInitialized = true;
+
+            UnselectAll(symbolTreeItemsSource);
+            var selectedNode = GetNodeSelectedByCaret(symbolTreeItemsSource);
+            if (selectedNode is not null)
+                SelectNode(selectedNode);
+
+            static void UnselectAll(ImmutableArray<DocumentSymbolViewModel> documentSymbolModels)
+            {
+                foreach (var documentSymbolModel in documentSymbolModels)
+                {
+                    documentSymbolModel.IsSelected = false;
+                    UnselectAll(documentSymbolModel.Children);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Sets the IsSelected field of a DocumentSymbolViewModel or one of its descendants to true.
+        /// </summary>
+        /// <remarks>
+        /// Assumes the caret position is in range of the given DocumentSymbolViewModel.
+        /// </remarks>
+        public static void SelectNode(DocumentSymbolViewModel symbol)
+        {
+            var selectedChildSymbol = GetNodeSelectedByCaret(symbol.Children);
+            if (selectedChildSymbol is not null)
+                SelectNode(selectedChildSymbol);
+            else
+                symbol.IsSelected = true;
+        }
+
+        /// <summary>
+        /// Returns a DocumentSymbolViewModel if the current caret position is in its range and null otherwise.
+        /// </summary>
+        public static DocumentSymbolViewModel? GetNodeSelectedByCaret(ImmutableArray<DocumentSymbolViewModel> symbolTreeItemsSource)
+        {
+            if (IsSnapshotInitialized)
+            {
+                foreach (var symbol in symbolTreeItemsSource)
+                {
+                    var oldStartPosition = LspSnapshot.GetLineFromLineNumber(symbol.StartPosition.Line).Start.Position + symbol.StartPosition.Character;
+                    var oldEndPosition = LspSnapshot.GetLineFromLineNumber(symbol.EndPosition.Line).Start.Position + symbol.EndPosition.Character;
+
+                    var currentStartPosition = new SnapshotPoint(CurrentSnapshot, oldStartPosition).Position;
+                    var currentEndPosition = new SnapshotPoint(CurrentSnapshot, oldEndPosition).Position;
```

Comment:
See earlier comments -- this isn't safe if code was deleted between snapshots; this needs to be using some other mapping.